### PR TITLE
bgp-cp: use CiliumNode object and support ClusterPool v1/v2 IPAM

### PR DESCRIPTION
**Pull request by ldelossa**
This PR refactors bits of the BGP Control Plane to understand Cilium's Cluster Pool IPAM.

When Cluster Pool IPAM is configured the BGP control plane will configure itself to parse the CiliumNode object's PodCIDR blocks instead of the Kubernetes native Node resource's. 

When Cluster Pool IPAM is not configured, the normal Kubernetes Node resource is parsed instead.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

```release-note
Move the BGP Control Plane to utilize CiliumNode objects. This enable support for IPAM driven PodCIDR announcements. 
```

### CODE REVIEW

**Comment by gandro** [comment]
> When Cilium's IPAM is not in use these modified fields will still contain the default PodCIDR range Kubernetes configured.

Note: This is not the case right now and I don't see any code in this PR which seems to change that?

File: pkg/bgpv1/agent/controller.go [hunk 1]
```diff
@@ -10,17 +10,15 @@ import (
...
-	"k8s.io/client-go/informers"
-	v1 "k8s.io/client-go/listers/core/v1"
 	"k8s.io/client-go/tools/cache"
 
 	"github.com/cilium/cilium/pkg/bgpv1"
 	"github.com/cilium/cilium/pkg/k8s"
+	v2api "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
 	v2alpha1api "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
 	"github.com/cilium/cilium/pkg/k8s/client/informers/externalversions"
+	v2 "github.com/cilium/cilium/pkg/k8s/client/listers/cilium.io/v2"
```
Comment (christarazi) [inline pkg/bgpv1/agent/controller.go]: Nit: 
```suggestion
	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
	v2alpha1api "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
	"github.com/cilium/cilium/pkg/k8s/client/informers/externalversions"
	listersv2 "github.com/cilium/cilium/pkg/k8s/client/listers/cilium.io/v2"
```

File: pkg/bgpv1/mock/mocks.go [hunk 2]
```diff
@@ -6,28 +6,28 @@ package mock
 import (
 	"context"
 
-	v1 "k8s.io/api/core/v1"
+	v2listers "github.com/cilium/cilium/pkg/k8s/client/listers/cilium.io/v2"
```
Comment (christarazi) [inline pkg/bgpv1/mock/mocks.go]: File is not `goimports`'d

File: pkg/bgpv1/agent/controller.go [hunk 3]
```diff
@@ -150,19 +145,27 @@ func NewController(ctx context.Context, rtMgr BGPRouterManager, opts ...Controll
 	clientset := k8s.CiliumClient()
 	factory := externalversions.NewSharedInformerFactory(clientset, 0)
 
-	selfTweakList := informers.WithTweakListOptions(func(lo *metav1.ListOptions) {
-		lo.FieldSelector = "metadata.name=" + nodetypes.GetName()
-	})
-	k8sfactory := informers.NewSharedInformerFactoryWithOptions(k8s.Client(), 0, selfTweakList)
+	// selfTweakList := informers.WithTweakListOptions(func(lo *metav1.ListOptions) {
```
Comment (christarazi) [inline pkg/bgpv1/agent/controller.go]: Remove commented out code?

**Review by christarazi — CHANGES_REQUESTED** [review]
Few minor comments, also reiterating Sebastian's concern as well.

**Comment by ldelossa** [comment]
I misunderstood my initial conversation with @gandro about this because of lack of coffee. now fully caffeinated we discussed how this should actually work. Placing into draft for now until its implemented.

**Comment by ldelossa** [comment]
@christarazi I'm resolving previous comments, as the entire history you commented on has been dumped.

**Comment by ldelossa** [comment]
@gandro let me know your opinion on this. Essentially, I just interface-out the details about grabbing NodeSpec info, so my code doesnt need to be concerned about the Cilium/KubernetesNode differences. 

The functions do a client lookup, however the lookup is backed by a SharedInformer which performs in-memory and indexed lookups, so we don't make API calls everytime we want to lookup the Node's labels for example.

File: pkg/bgpv1/agent/controller.go [hunk 4]
```diff
@@ -143,49 +177,73 @@ type ControllerOpt func(*Controller)
...
+	case ipamOption.IPAMClusterPoolV2:
+		var err error
+		nodeSpecer, err = configureForClusterPoolIPAM(factory, sig, stop)
+		if err != nil {
+			return nil, fmt.Errorf("failed to configure listers and informers for ClusterPool IPAM: %w", err)
+		}
+	case ipamOption.IPAMKubernetes:
+		var err error
+		selfTweakList := informers.WithTweakListOptions(func(lo *metav1.ListOptions) {
+			lo.FieldSelector = "metadata.name=" + nodetypes.GetName()
```
Comment (gandro) [inline pkg/bgpv1/agent/controller.go:212]: Why do you only filter for the node name here, and not also in the ClusterPool case?

File: pkg/bgpv1/agent/nodespecer.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	}
+	return n.Labels, nil
+}
+
+func (s *ciliumNodeSpecer) PodCIDRs() ([]string, error) {
+	n, err := s.NodeLister.Get(s.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get node %v: %v", s.Name, err)
+	}
+	if n.Spec.IPAM.PodCIDRs != nil {
```
Comment (gandro) [inline pkg/bgpv1/agent/nodespecer.go:87]:
```suggestion
	if len(n.Spec.IPAM.PodCIDRs) > 0 {
```

File: pkg/bgpv1/agent/nodespecer.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	}
+	return n.Labels, nil
+}
+
+func (s *kubernetesNodeSpecer) PodCIDRs() ([]string, error) {
+	n, err := s.NodeLister.Get(s.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get node %v: %v", s.Name, err)
+	}
+	if n.Spec.PodCIDRs != nil {
```
Comment (gandro) [inline pkg/bgpv1/agent/nodespecer.go:52]:
```suggestion
	if len(n.Spec.PodCIDRs) > 0 {
```

File: pkg/bgpv1/agent/controller.go [hunk 7]
```diff
@@ -143,49 +177,73 @@ type ControllerOpt func(*Controller)
...
+	// setup is dictate by the type of IPAM being used. If Kubernetes IPAM is
+	// being used the BGP Control Plane must watch and retrieve PodCIDR ranges
+	// from the corev1 Kubernetes Node resource.
+	//
+	// if the ClusterPool (both v1 and v2) IPAM mode is in use, the control plane
+	// must watch and retrieve PodCIDR ranges from Cilium's v2 CiliumNode resource.
+	switch option.Config.IPAM {
+	case ipamOption.IPAMClusterPool:
+		fallthrough
+	case ipamOption.IPAMClusterPoolV2:
```
Comment (gandro) [inline pkg/bgpv1/agent/controller.go]: tiny nit (feel free to ignore)
```suggestion
	case ipamOption.IPAMClusterPool, ipamOption.IPAMClusterPoolV2:
```

**Review by gandro — COMMENTED** [review]
While I'm not familiar with the new bgp-cp, the IPAM part looks good to me, only one question that popped up (the rest are nits).

File: pkg/bgpv1/agent/nodespecer.go (hunk 5 above)
Comment (ldelossa) [reply pkg/bgpv1/agent/nodespecer.go:87]: Won't the zero value for n.Spec.IPAM.PodCIDRs be nil? If it is, we just return the empty []string{}.

Comment (ldelossa) [reply pkg/bgpv1/agent/nodespecer.go:87]: Won't the zero value for n.Spec.IPAM.PodCIDRs be nil? If it is, we just return the empty []string{}.

Comment (ldelossa) [reply pkg/bgpv1/agent/nodespecer.go:87]: Won't the zero value for n.Spec.IPAM.PodCIDRs be nil? If it is, we just return the empty []string{}.

Comment (gandro) [reply pkg/bgpv1/agent/nodespecer.go:87]: True. Even if `PodCIDRs` is `[]string{}` (and not nil), we still return the correct value. Nevermind my nit then. Marking as resolved.

File: pkg/bgpv1/agent/controller.go (hunk 4 above)
Comment (ldelossa) [reply pkg/bgpv1/agent/controller.go:212]: Ah, good point. 

Because I use the other factory to **also**  list BGPPeeringPolicies, I don't think I can place the node filter on it, or else I'd limit the listing resources by that "metadata.name" field, which the BGPPeeringPolicies do not have. Hmm. is there a way to say "Filter this resource, by this field" and leave other resources unfiltered?

File: pkg/bgpv1/agent/controller.go (hunk 7 above)
Comment (ldelossa) [reply pkg/bgpv1/agent/controller.go]: oh nice! forgot about that syntax, I like this more. :+1:

File: pkg/bgpv1/agent/controller.go (hunk 4 above)
Comment (ldelossa) [reply pkg/bgpv1/agent/controller.go:212]: @gandro 
Ok, so I opt'd to filter the CiliumNode factory by metadata.name as well. 
I dug a bit into the code and the in-memory caches are not created until a constructor for the informers are called. 

In a subsequent PR I actually want to get rid of both the CiliumNode and Node informers in favor of grabbing the stores from the watchers.go file and using those to list from. It's a larger change tho and I'd like to keep that to its own PR if possible.

**Comment by joestringer** [comment]
ConformanceKind1.19 failure is https://github.com/cilium/cilium/issues/20034

File: pkg/bgpv1/agent/controller.go [hunk 8]
```diff
@@ -198,7 +256,7 @@ func NewController(ctx context.Context, rtMgr BGPRouterManager, opts ...Controll
 //
 // A cancel of the provided ctx will kill the control loop along with the running
 // informers.
-func (c *Controller) Run(ctx context.Context) {
+func (c *Controller) Run(ctx context.Context, stop chan struct{}) {
```
Comment (gandro) [inline pkg/bgpv1/agent/controller.go:260]: The new stop channel argument seems unused?

Comment (ldelossa) [reply pkg/bgpv1/agent/controller.go:260]: Believe its passed @ https://github.com/cilium/cilium/pull/19872/files#diff-663c8aba9492de183ebe045c8bf425d6a6a645970cd2a201a7073e9b21ffe8adR246

Look okay?

Comment (gandro) [reply pkg/bgpv1/agent/controller.go:260]: The link doesn't work for me, but yes, sorry, looking at the full file it is used here:
https://github.com/cilium/cilium/blob/6687a6f3887eb7f6cd660d2d00cb57c1ba40f253/pkg/bgpv1/agent/controller.go#L277

File: pkg/bgpv1/agent/controller.go [hunk 9]
```diff
@@ -1,4 +1,3 @@
-// SPDX-License-Identifier: Apache-2.0
```
Comment (gandro) [inline pkg/bgpv1/agent/controller.go]: Why is this removed?

File: pkg/bgpv1/agent/controller.go (hunk 8 above)
Comment (ldelossa) [reply pkg/bgpv1/agent/controller.go:260]: Cool. Marking resolved

**Review by gandro — COMMENTED** [review]
From what I know about the logic, this looks good to me. One minor question

File: pkg/bgpv1/agent/controller.go (hunk 9 above)
Comment (ldelossa) [reply pkg/bgpv1/agent/controller.go]: good catch, unintentional.

**Review by christarazi — APPROVED** [review]

**Review by gandro — APPROVED** [review]

**Comment by ldelossa** [comment]
GKE-stable failure unrelated. Marking as ready to merge.
